jobs/gorouter/spec

```diff
@@ -518,6 +518,25 @@ properties:
     description: "The number of file descriptors a router can have open at one time"
     default: 100000
 
+  router.enable_verify_client_certificate_metadata:
+    description: |
+        Enable additional client certificate verification via verify_client_certificate_metadata (see below).
+    default: false
+  router.verify_client_certificate_metadata:
+    description: |
+      Additional client certificate verification which limits the allowed client certificate for given to a signing CA (identified by its subject) to the certificates with subjects provided in the list of valid subjects. Each list entry contains a ca_subject with a coresponding list of valid subjects.
```

Comment:
Another thing that's a bit confusing here:
- We provide a list of `ca_subject` entries, coupled with a list of `valid_subjects`

Usually when speaking about x509 certificates, there is an "issuer" (the CA) and a "subject" (the thing that is signed by the certificate). What may be confusing here is that the "issuer" may as well be a "subject" one step above in the chain where the previous certificate was signed by the preceding CA. However, I would like to make it clear that for the trust check we apply here, we will only ever look at leaf certificates, which will only have one "issuer" DN and one "subject" DN.

So to me it would make more sense to speak of a list of `issuer` entries coupled with a list of "subject" entries.

Reply:
The whole chain is inspected, where the certificate's issuer is not necessarily the root CA certificate defined in the `ca_subject`.

Example:
- root CA, issuer "CN=root CA", subject "CN=root CA"
- ICA, issuer "CN=root CA", subject "CN=Intermediate CA"
- Client Cert, issuer "CN=intermediate CA", subject "CN=Client, Inc."

You'd define:
```yaml
- ca_subject:
    common_name: root CA
  valid_subjects:
    - common_name: Client, Inc.
```

You _could_ also define a ca_subject with the intermediate CA:
```yaml
- ca_subject:
    common_name: intermediate CA
  valid_subjects:
    - common_name: Client, Inc.
```

Ulitmately jus the DN of one of the CA certificates (root or intermediate) in the certificate chain must match the given subject. It's not necessarily the client certificate's issuer.

Reply:
my point remains. to me it would make equal sense to use this notation:
```
- issuer:
    common_name: root CA
  valid_subjects:
    - common_name: Client, Inc.
```
this means "if anywhere in the chain we have `root CA` as the issuer, then we will accept `Client, Inc.` as valid subjects of a **leaf** certificate". (we might even rename `valid_subjects` to `valid_leaf_subjects` to make it clearer...

the same goes for the intermediate:
```
- issuer:
    common_name: intermediate CA
  valid_subjects:
    - common_name: Client, Inc.
```
here it becomes even clearer. we mean to say: "if `intermediate CA` is the issuer, we will accept `Client, Inc.` as valid subjects for any **leaf** certificate"

Reply:
ok, fair. `ca_subject` exposes an implementation detail, i.e. that we check the CA certificates' subjects in the chain.

Even clearer could be the following?
```yaml
- issuer_in_chain:
    common_name: root CA
  valid_cert_subjects:
    - common_name: Client, Inc.
```

WDYT?

Reply:
Yes! I really like it much better 😁